Sources/_MatchingEngine/Regex/Parse/CaptureStructure.swift

```diff
@@ -0,0 +1,196 @@
+// A tree representing the type of some captures.
+public enum CaptureStructure: Equatable {
+  case atom(name: String? = nil)
+  indirect case array(CaptureStructure)
+  indirect case optional(CaptureStructure)
+  indirect case tuple([CaptureStructure])
+
+  public static func tuple(_ children: CaptureStructure...) -> Self {
+    tuple(children)
+  }
+
+  public static var empty: Self {
+    .tuple([])
+  }
+}
+
+// MARK: - Combination and transformation
+
+extension CaptureStructure {
+  /// Returns a capture structure by concatenating any tuples in `self` and
+  /// `other`.
+  func concatenating(with other: CaptureStructure) -> CaptureStructure {
+    switch (self, other) {
+    // (T...) + (U...) ==> (T..., U...)
+    case let (.tuple(lhs), .tuple(rhs)):
+      return .tuple(lhs + rhs)
+    // T + () ==> T
+    case (_, .tuple(let rhs)) where rhs.isEmpty:
+      return self
+    // () + T ==> T
+    case (.tuple(let lhs), _) where lhs.isEmpty:
+      return other
+    // (T...) + U ==> (T..., U)
+    case let (.tuple(lhs), _):
+      return .tuple(lhs + [other])
+    // T + (U...) ==> (T, U...)
+    case let (_, .tuple(rhs)):
+      return .tuple([self] + rhs)
+    // T + U ==> (T, U)
+    default:
+      return .tuple([self, other])
+    }
+  }
+
+  /// Returns a capture structure by transforming any tuple element to an
+  /// optional thereof. This is used specifically for forming the capture type
+  /// of an alternation.
+  func alternating() -> CaptureStructure {
+    // (T...) ==> (T?...)
+    if case .tuple(let children) = self {
+      return .tuple(children.map(CaptureStructure.optional))
+    }
+    // T ==> T
+    return self
+  }
+
+  /// Returns a capture structure by transforming any tuple element of `self`
+  /// or transforming `self` directly if it is not a tuple.
+  func compactMap(
+    _ transform: (CaptureStructure) -> CaptureStructure
+  ) -> CaptureStructure {
```

Comment:
IMO this seems closer to a regular `map` than a `compactMap`, what do you think?